query/physicalplan/sampler_test.go

```diff
@@ -194,3 +209,127 @@ func Benchmark_Sampler(b *testing.B) {
 		})
 	}
 }
+
+func Test_Sampler_Materialize(t *testing.T) {
+	ctx := context.Background()
+	allocator := memory.NewCheckedAllocator(memory.NewGoAllocator())
+	t.Cleanup(func() {
+		allocator.AssertSize(t, 0)
+	})
+	s := NewReservoirSampler(10, 10_000, allocator)
+	s.SetNext(&TestPlan{})
+
+	schema := arrow.NewSchema([]arrow.Field{
+		{Name: "a", Type: arrow.PrimitiveTypes.Int64},
+	}, nil)
+	bldr := array.NewRecordBuilder(allocator, schema)
+	t.Cleanup(bldr.Release)
+
+	for i := 0; i < 10; i++ {
+		for j := 0; j < 10; j++ {
+			bldr.Field(0).(*array.Int64Builder).Append(int64((i * 10) + j))
+		}
+		r := bldr.NewRecord()
+		t.Cleanup(r.Release)
+		require.NoError(t, s.Callback(ctx, r))
+	}
+
+	// Create a new schema for records
+	schema = arrow.NewSchema([]arrow.Field{
+		{Name: "a", Type: arrow.PrimitiveTypes.Int64},
+		{Name: "b", Type: arrow.PrimitiveTypes.Int64},
+	}, nil)
+	bldr = array.NewRecordBuilder(allocator, schema)
+	t.Cleanup(bldr.Release)
+
+	for i := 0; i < 10; i++ {
+		for j := 0; j < 10; j++ {
+			bldr.Field(0).(*array.Int64Builder).Append(int64((i * 10) + j))
+			bldr.Field(1).(*array.Int64Builder).Append(int64((i * 10) + j))
+		}
+		r := bldr.NewRecord()
+		t.Cleanup(r.Release)
+		require.NoError(t, s.Callback(ctx, r))
+	}
+
+	require.NoError(t, s.materialize(allocator))
+	s.Close()
+	require.Zero(t, s.sizeInBytes)
+}
+
+type AccountingAllocator struct {
+	allocator *memory.CheckedAllocator
+	maxUsed   int
+}
+
+func (a *AccountingAllocator) Allocate(size int) []byte {
+	b := a.allocator.Allocate(size)
+	if current := a.allocator.CurrentAlloc(); current > a.maxUsed {
+		a.maxUsed = current
+	}
+	return b
+}
+
+func (a *AccountingAllocator) Reallocate(size int, data []byte) []byte {
+	b := a.allocator.Reallocate(size, data)
+	if current := a.allocator.CurrentAlloc(); current > a.maxUsed {
+		a.maxUsed = current
+	}
+	return b
+}
+
+func (a *AccountingAllocator) Free(data []byte) {
+	a.allocator.Free(data)
+	if current := a.allocator.CurrentAlloc(); current > a.maxUsed {
+		a.maxUsed = current
+	}
+}
+
+func Test_Sampler_MaxSizeAllocation(t *testing.T) {
+	ctx := context.Background()
+	allocator := &AccountingAllocator{
+		allocator: memory.NewCheckedAllocator(memory.NewGoAllocator()),
+	}
+	t.Cleanup(func() {
+		require.Equal(t, allocator.maxUsed, 1024) // Expect the most we allocated was 1024 bytes during materialization
```

Comment:
It gets weird mixing cleanup and defer. Because we cleanup the releases we needed this to execute last. If I use defer it executes before the other cleanups.

Reply:
But given defer is LIFO why not just defer everything you've put in cleanup functions in this test?